syntax = "proto3";

service Movie {
    rpc GetMovieByID(MovieID) returns (MovieData) {}
    rpc GetListMovies(Empty) returns (stream MovieData) {}
    rpc GetMovieByTitle(MovieTitle) returns (MovieData) {}
    rpc GetMovieByDirector(MovieDirector) returns (stream MovieData) {}
    rpc AddMovie(MovieData) returns (stream MovieData) {}
    rpc DeleteMovieByID(MovieID) returns (Empty) {}
    rpc UpdateMovieRate(MovieIDAndRate) returns (MovieData) {}
    rpc DeleteMovieByRate(MovieRate) returns (Empty) {}
}

message MovieID {
    string id = 1;
}

message MovieTitle {
    string title = 1;
}

message MovieDirector {
    string director = 1;
}

message MovieData {
    string title = 1;
    float rating = 2;
    string director = 3;
    string id = 4;
}

message MovieRate {
    float rating = 1;
}

message MovieIDAndRate {
    MovieID id = 1;
    MovieRate rating = 2;
}

message Empty {
}